codec-dns/src/main/java/io/netty/handler/codec/dns/DefaultDnsRecordDecoder.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.dns;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.handler.codec.CorruptedFrameException;
+import io.netty.util.CharsetUtil;
+
+public class DefaultDnsRecordDecoder implements DnsRecordDecoder {
+
+    @Override
+    public DnsQuestion decodeQuestion(ByteBuf in) throws Exception {
+        String name = decodeName(in);
+        DnsRecordType type = DnsRecordType.valueOf(in.readUnsignedShort());
+        DnsRecordClass qClass = DnsRecordClass.valueOf(in.readUnsignedShort());
+        return new DefaultDnsQuestion(name, type, qClass);
+    }
+
+    @Override
+    public <T extends DnsRecord> T decodeRecord(ByteBuf in) throws Exception {
+        final String name = decodeName(in);
+        final DnsRecordType type = DnsRecordType.valueOf(in.readUnsignedShort());
+        final DnsRecordClass aClass = DnsRecordClass.valueOf(in.readUnsignedShort());
+        final long ttl = in.readUnsignedInt();
+        final int length = in.readUnsignedShort();
+        final int offset = in.readerIndex();
+
+        @SuppressWarnings("unchecked")
+        T record = (T) decodeRecord(name, type, aClass, ttl, in, offset, length);
```

Comment:
It's re @Scottmitch question can this be done more safely.  

It just lets you not `SuppressWarnings("unchecked")` and potentially throw with a more meaningful error message than the default CCE message.  And slightly more future-proof in general, since the code actually knows the expected return type and can react to it (if you later widened the type of `T` callers could still get what they expect, or certain types could be special-cased).